### DISCUSSION TITLE
Seq1 migration issue?

### DISCUSSION THREAD

**Discussion by StefanBertels** [louthy/language-ext#931] [General]
Release https://github.com/louthy/language-ext/releases/tag/v4.0.2 changed static constructor function `Seq1` to `Seq`.

I fully suport this decision as `Seq1` always felt like "wrong" or like a workaround.

But I just want to point out some issue that might occur when upgrading existing code to this new release.
IMO a good way to migrate is via compile time errors (maximum safety).
Of course there are other options like `[Obsolete]` which might be helpful.

I guess there are relevant use cases out there that might *not* result in compile time errors. This might produce bugs if users don't take special care (i.e. don't read release notes or maybe even ignore warnings about obsolete functions).

Here's some test that will fail on 4.0.2 (but not 3.4.15).

```
[Fact]
void Bad()
{
    var data = LanguageExt.Prelude.Seq<int>();
    var x = LanguageExt.Prelude.Seq(data); // breaking change
    Assert.True(x.IsEmpty);
}
 ```
 
One idea for a safe migration path: Create some 3.4.16 (stable) release and another 3.x beta release where single parameter `Seq(x)` function is missing completely (forcing migration to `toSeq` when updating to this intermediate version).

There are other options using `[Obsolete]` and/or slow migration releases but IMO they are not appropriate for an OSS library like this (and maybe not safe).

**Comment by louthy** [comment]
I think this is a fair point.  What I might do is just remove the `Seq` constructor function, and resurrect `Seq1`.  `Seq1` was the thing that bothered me the least about this, it was the fact that `Seq` was used as both a constructor function and a conversion function.

> There are other options using [Obsolete] and/or slow migration releases but IMO they are not appropriate for an OSS library like this (and maybe not safe)

I think a migration approach would be simpler to be honest. If I migrate the conversion functions now (to `toSeq`), and keep `Seq1`, that will force a migration of the conversion functions.  Then 6 months or so I can convert `Seq1`, that should give everyone enough time to get to v4 and be ready for the switch.  For those that are really slow to migrate, there's a risk, but if I tag the releases and have it documented, then they can stage their upgrade in their own time.

**Reply by StefanBertels** [reply]
Totally agree.
